File: tests/fuzz/tools/analyze-failures.py
#!/usr/bin/env python3
"""Analyze failure patterns from fuzz test output."""

import re
import sys
from collections import Counter, defaultdict

def parse_test_output(filename="/tmp/fuzz-output.txt"):
    """Parse test output and extract failure patterns."""
    with open(filename, 'r') as f:
        content = f.read()

    # Extract all test failures
    test_pattern = r'Test #(\d+):\s+Mismatch at position \d+: Prettier="([^"]+)", RustyWind="([^"]+)"'
    failures = re.findall(test_pattern, content)

    return failures

def categorize_failure(prettier_class, rustywind_class):
    """Categorize the type of failure."""
    categories = []

    # Count variants
    prettier_variants = prettier_class.count(':')
    rustywind_variants = rustywind_class.count(':')

    # Check for common patterns
    if 'dark:' in prettier_class or 'dark:' in rustywind_class:
        categories.append('dark-variant')

    if 'placeholder:' in prettier_class or 'placeholder:' in rustywind_class:
        categories.append('placeholder-variant')

    if 'dark:placeholder:' in prettier_class or 'dark:placeholder:' in rustywind_class:
        categories.append('dark-placeholder-stack')

    if 'before:' in prettier_class or 'before:' in rustywind_class or 'after:' in prettier_class or 'after:' in rustywind_class:
        categories.append('pseudo-element')

    if 'group:' in prettier_class or 'group:' in rustywind_class or 'peer:' in prettier_class or 'peer:' in rustywind_class:
        categories.append('group-peer-base')

    if 'group-' in prettier_class or 'group-' in rustywind_class or 'peer-' in prettier_class or 'peer-' in rustywind_class:
        categories.append('group-peer-compound')

    if prettier_variants >= 2 or rustywind_variants >= 2:
        categories.append('multi-variant')
        if prettier_variants >= 3 or rustywind_variants >= 3:
            categories.append('triple-variant')

    # Check for specific patterns
    if re.search(r'(first|last|even|odd|only):', prettier_class) or re.search(r'(first|last|even|odd|only):', rustywind_class):
        categories.append('positional-pseudo')

    if re.search(r'(md|lg|xl|2xl|sm):', prettier_class) or re.search(r'(md|lg|xl|2xl|sm):', rustywind_class):
        categories.append('responsive')

    if '[' in prettier_class or '[' in rustywind_class:
        categories.append('arbitrary-value')

    if '/' in prettier_class or '/' in rustywind_class:
        categories.append('opacity-syntax')

    return categories if categories else ['other']

def main():
    failures = parse_test_output()

    print(f"Found {len(failures)} failures\n")
    print("=" * 80)

    # Categorize all failures
    category_counts = Counter()
    category_examples = defaultdict(list)

    for test_num, prettier, rustywind in failures:
        cats = categorize_failure(prettier, rustywind)
        for cat in cats:
            category_counts[cat] += 1
            if len(category_examples[cat]) < 3:
                category_examples[cat].append((prettier, rustywind))

    # Print category breakdown
    print("\n📊 FAILURE CATEGORIES\n")
    print("─" * 80)
    for cat, count in sorted(category_counts.items(), key=lambda x: -x[1]):
        pct = (count / len(failures)) * 100
        print(f"{cat:25s} {count:3d} failures ({pct:5.1f}%)")
    print("─" * 80)

    # Print examples for each category
    print("\n📋 EXAMPLE FAILURES BY CATEGORY\n")
    print("=" * 80)

    for cat in sorted(category_counts.keys(), key=lambda x: -category_counts[x]):
        print(f"\n{cat.upper().replace('-', ' ')} ({category_counts[cat]} failures):")
        print("─" * 80)
        for prettier, rustywind in category_examples[cat][:3]:
            print(f"  Prettier: {prettier}")
            print(f"  RustyWind: {rustywind}")
            print()

    # Special analysis for most common pattern
    dark_placeholder_count = sum(1 for _, p, r in failures if 'dark:placeholder:' in p or 'dark:placeholder:' in r)
    print("\n" + "=" * 80)
    print(f"\n⚠️  MOST COMMON PATTERN: dark:placeholder: ({dark_placeholder_count} occurrences)")
    print("─" * 80)
    print("This double-stacked variant appears in {:.1f}% of all failures".format(
        (dark_placeholder_count / len(failures)) * 100
    ))

    # Show dark:placeholder examples
    print("\nExamples where dark:placeholder: classes sort incorrectly:")
    dark_placeholder_examples = [(p, r) for _, p, r in failures if 'dark:placeholder:' in p or 'dark:placeholder:' in r][:5]
    for i, (p, r) in enumerate(dark_placeholder_examples, 1):
        print(f"\n{i}. Prettier: {p}")
        print(f"   RustyWind: {r}")

if __name__ == "__main__":
    main()

File: tests/fuzz/tools/analyze-failures.js
/**
 * Analyze failure patterns from multi-seed results
 */
import { readFile, writeFile } from 'fs/promises';

const RESULTS_FILE = 'multi-seed-results.json';

/**
 * Categorize a failure pattern by the types of classes involved
 */
function categorizeFailure(pattern) {
  const categories = [];

  // Extract the class names from the mismatch message
  const match = pattern.match(/Prettier="([^"]+)", RustyWind="([^"]+)"/);
  if (!match) return ['unknown'];

  const prettierClass = match[1];
  const rustywindClass = match[2];

  // Categorize by class type patterns
  if (prettierClass.includes('outline') || rustywindClass.includes('outline')) {
    categories.push('outline');
  }
  if (prettierClass.includes('transition') || rustywindClass.includes('transition')) {
    categories.push('transition');
  }
  if (prettierClass.includes('duration') || rustywindClass.includes('duration')) {
    categories.push('duration');
  }
  if (prettierClass.includes('delay') || rustywindClass.includes('delay')) {
    categories.push('delay');
  }
  if (prettierClass.includes('rounded') || rustywindClass.includes('rounded')) {
    categories.push('rounded');
  }
  if (prettierClass.includes('bg-') || rustywindClass.includes('bg-')) {
    categories.push('background');
  }
  if (prettierClass.includes('space-') || rustywindClass.includes('space-')) {
    categories.push('spacing');
  }
  if (prettierClass.includes('gap-') || rustywindClass.includes('gap-')) {
    categories.push('gap');
  }
  if (prettierClass.includes('divide-') || rustywindClass.includes('divide-')) {
    categories.push('divide');
  }
  if (prettierClass.includes('ring-') || rustywindClass.includes('ring-')) {
    categories.push('ring');
  }
  if (prettierClass.includes('shadow') || rustywindClass.includes('shadow')) {
    categories.push('shadow');
  }
  if (prettierClass.includes('border-') || rustywindClass.includes('border-')) {
    categories.push('border');
  }
  if (prettierClass.includes('-rotate') || rustywindClass.includes('-rotate')) {
    categories.push('rotation');
  }
  if (prettierClass.includes('will-change') || rustywindClass.includes('will-change')) {
    categories.push('will-change');
  }

  // If no specific category found, try to identify generic category
  if (categories.length === 0) {
    categories.push('other');
  }

  return categories;
}

/**
 * Extract the actual classes from failure details
 */
function extractClasses(example) {
  const details = example.details;
  if (!details || !details.lines) return null;

  const original = details.lines.find(l => l.includes('Original:'));
  const prettier = details.lines.find(l => l.includes('Prettier:'));
  const rustywind = details.lines.find(l => l.includes('RustyWind:'));

  if (!original || !prettier || !rustywind) return null;

  // Extract arrays from lines like "  Original:  [class1, class2, ...]"
  const extractArray = (line) => {
    const match = line.match(/\[(.*)\]/);
    if (!match) return [];
    return match[1].split(',').map(s => s.trim());
  };

  return {
    original: extractArray(original),
    prettier: extractArray(prettier),
    rustywind: extractArray(rustywind)
  };
}

/**
 * Main analysis function
 */
async function analyzeFailures() {
  console.log('📊 Analyzing failure patterns...\n');

  const data = JSON.parse(await readFile(RESULTS_FILE, 'utf-8'));

  console.log('Summary from results:');
  console.log(`  Total rounds: ${data.summary.totalRounds}`);
  console.log(`  Total tests: ${data.summary.totalTests}`);
  console.log(`  Total passed: ${data.summary.totalPassed}`);
  console.log(`  Total failed: ${data.summary.totalFailed}`);
  console.log(`  Unique patterns: ${data.summary.uniqueFailurePatterns}\n`);

  // Categorize all failures
  const categoryMap = new Map(); // category -> list of patterns
  const categoryExamples = new Map(); // category -> example with classes

  for (const failurePattern of data.failurePatterns) {
    const categories = categorizeFailure(failurePattern.pattern);

    // Try to get a good example with full class details
    const exampleWithClasses = failurePattern.examples.find(ex => {
      const classes = extractClasses(ex);
      return classes !== null;
    });

    for (const category of categories) {
      if (!categoryMap.has(category)) {
        categoryMap.set(category, []);
      }
      categoryMap.get(category).push({
        pattern: failurePattern.pattern,
        occurrences: failurePattern.occurrenceCount,
        example: failurePattern.examples[0],
        exampleWithClasses
      });

      // Store first good example for each category
      if (!categoryExamples.has(category) && exampleWithClasses) {
        categoryExamples.set(category, exampleWithClasses);
      }
    }
  }

  // Sort categories by frequency
  const sortedCategories = Array.from(categoryMap.entries())
    .map(([category, patterns]) => ({
      category,
      patternCount: patterns.length,
      totalOccurrences: patterns.reduce((sum, p) => sum + p.occurrences, 0),
      patterns
    }))
    .sort((a, b) => b.totalOccurrences - a.totalOccurrences);

  console.log('📋 Failure Categories (sorted by frequency):\n');

  for (const { category, patternCount, totalOccurrences, patterns } of sortedCategories) {
    console.log(`${category.toUpperCase()}:`);
    console.log(`  Unique patterns: ${patternCount}`);
    console.log(`  Total occurrences: ${totalOccurrences}`);
    console.log(`  Top 3 patterns:`);

    const topPatterns = patterns
      .sort((a, b) => b.occurrences - a.occurrences)
      .slice(0, 3);

    for (const p of topPatterns) {
      console.log(`    - ${p.pattern} (${p.occurrences}x)`);
    }
    console.log('');
  }

  // Generate detailed category report
  const report = {
    summary: data.summary,
    categories: sortedCategories.map(({ category, patternCount, totalOccurrences, patterns }) => ({
      category,
      patternCount,
      totalOccurrences,
      topPatterns: patterns
        .sort((a, b) => b.occurrences - a.occurrences)
        .slice(0, 5)
        .map(p => ({
          pattern: p.pattern,
          occurrences: p.occurrences,
          seed: p.example.seed,
          testNumber: p.example.testNumber,
          classes: extractClasses(p.exampleWithClasses || p.example)
        }))
    }))
  };

  await writeFile('failure-analysis.json', JSON.stringify(report, null, 2));
  console.log('💾 Detailed analysis saved to: failure-analysis.json\n');

  // Print recommendations
  console.log('📝 Recommendations for static tests:\n');
  console.log('Based on the analysis, we should create static tests for these categories:');
  for (const { category, totalOccurrences } of sortedCategories.slice(0, 10)) {
    console.log(`  - ${category} (${totalOccurrences} failures)`);
  }
}

analyzeFailures().catch(error => {
  console.error('Error:', error);
  process.exit(1);
});
